### ISSUE TITLE
`gl_VertexIndex` is an array instead of an integer

### ISSUE THREAD

**Issue by cab** [gfx-rs/naga#804]
When using `gl_VertexIndex` in a GLSL shader compiled to SPIRV, it is validated as an array instead of an integer:

### shader code
```glsl
switch (gl_VertexIndex) {
}
```

### error
```
wgpu error: Validation Error

Caused by:
    In Device::create_shader_module
      note: label = `shader/vertex.spv`
    Function [1] 'main' is invalid
    The `switch` value [35] is not an integer scalar
```

### logs
`RUST_LOG=warn` shows:
```
[2021-04-29T21:30:10Z WARN  gfx_backend_vulkan] Unable to find layer: VK_LAYER_KHRONOS_validation
INTEL-MESA: warning: Vulkan is not yet fully supported on gen12
INTEL-MESA: warning: Performance support disabled, consider sysctl dev.i915.perf_stream_paranoid=0
[2021-04-29T21:30:10Z WARN  naga::front::spv] Treating VertexIndex as unsigned
```

### workaround
```glsl
uint gli = uint(gl_VertexIndex);

switch (int(gli)) {
}
```

### links

[Matrix discussion](https://matrix.to/#/!FZyQrssSlHEZqrYcOb:matrix.org/$0o8av9v92-KEqDZyWpzX4fW3S9WyV5sg24wJuIcahfI?via=matrix.org)

**Comment by kvark** [comment]
Please specify what Naga version was used here (perhaps, looking at your "Cargo.lock").
I tested the following program on master, and it's all good:
```glsl
#version 440
void main()
{
    switch(gl_VertexIndex) {
            case 1: gl_Position = vec4(0.0); break;
            default: gl_Position = vec4(1.0); break;
    }
}
```

**Comment by cab** [comment]
ah, it was `0.3.1`, so maybe this is already fixed.

the full shader was
```glsl
#version 450

layout(set = 0, binding = 0) uniform Transform {
    mat4 transform;
};

layout(location = 0) in vec3 left_top;
layout(location = 1) in vec2 right_bottom;
layout(location = 2) in vec2 tex_left_top;
layout(location = 3) in vec2 tex_right_bottom;
layout(location = 4) in vec4 color;

layout(location = 0) out vec2 f_tex_pos;
layout(location = 1) out vec4 f_color;

// generate positional data based on vertex ID
void main() {
    vec2 pos = vec2(0.0);
    float left = left_top.x;
    float right = right_bottom.x;
    float top = left_top.y;
    float bottom = right_bottom.y;

    switch (gl_VertexIndex) {
        case 0:
            pos = vec2(left, top);
            f_tex_pos = tex_left_top;
            break;

        case 1:
            pos = vec2(right, top);
            f_tex_pos = vec2(tex_right_bottom.x, tex_left_top.y);
            break;

        case 2:
            pos = vec2(left, bottom);
            f_tex_pos = vec2(tex_left_top.x, tex_right_bottom.y);
            break;

        case 3:
            pos = vec2(right, bottom);
            f_tex_pos = tex_right_bottom;
            break;
    }

    f_color = color;
    gl_Position = transform * vec4(pos, left_top.z, 1.0);
}
```

**Comment by kvark** [comment]
Ok, this has been fixed since 0.3.1 :)

**Comment by kvark** [comment]
No problem! Sorry for asking you to file this without querying your version first :)
